File: apps/web/src/data/ordering.ts
// Code-point order. `localeCompare` depends on the runtime's collation, and a
// server that serves this contract in another language cannot be relied on to
// reproduce it. Every ordering the contract promises is defined here so both
// implementations can agree exactly.
export function compareText(left: string, right: string): number {
  if (left < right) return -1;
  return left > right ? 1 : 0;
}

export function compareNames(left: string, right: string): number {
  return compareText(left.toLowerCase(), right.toLowerCase()) || compareText(left, right);
}

File: crates/domain/src/ordering.rs
use std::cmp::Ordering;

/// Code-point order, which `apps/web/src/data/ordering.ts` also defines. The
/// browser's `localeCompare` depends on its collation, so an ordering the
/// contract promises is defined without it and both sides can agree exactly.
pub fn compare_text(left: &str, right: &str) -> Ordering {
    left.cmp(right)
}

pub fn compare_names(left: &str, right: &str) -> Ordering {
    left.to_lowercase()
        .cmp(&right.to_lowercase())
        .then_with(|| left.cmp(right))
}
